presto-main/src/main/java/com/facebook/presto/operator/AggregationOperator.java

```diff
@@ -218,42 +272,119 @@ private Aggregator(AccumulatorFactory accumulatorFactory, Step step)
                 intermediateChannel = accumulatorFactory.getInputChannels().get(0);
                 aggregation = accumulatorFactory.createIntermediateAccumulator();
             }
+
+            if (orderByChannels.isEmpty()) {
+                checkArgument(orderings.isEmpty(), "orderByChannels is empty but ordering is not");
+                this.accumulator = aggregation;
+            }
+            else {
+                checkState(pagesIndexFactory.isPresent(), "No pagesIndexFactory to process ordering");
+                this.accumulator = new OrderingAccumulator(aggregation, orderBySourceTypes, orderByChannels, orderings, pagesIndexFactory.get());
+            }
+
             this.step = step;
         }
 
         public Type getType()
         {
             if (step.isOutputPartial()) {
-                return aggregation.getIntermediateType();
+                return accumulator.getIntermediateType();
             }
             else {
-                return aggregation.getFinalType();
+                return accumulator.getFinalType();
             }
         }
 
         public void processPage(Page page)
         {
             if (step.isInputRaw()) {
-                aggregation.addInput(page);
+                accumulator.addInput(page);
             }
             else {
-                aggregation.addIntermediate(page.getBlock(intermediateChannel));
+                accumulator.addIntermediate(page.getBlock(intermediateChannel));
             }
         }
 
         public void evaluate(BlockBuilder blockBuilder)
         {
             if (step.isOutputPartial()) {
-                aggregation.evaluateIntermediate(blockBuilder);
+                accumulator.evaluateIntermediate(blockBuilder);
             }
             else {
-                aggregation.evaluateFinal(blockBuilder);
+                accumulator.evaluateFinal(blockBuilder);
             }
         }
 
         public long getEstimatedSize()
         {
-            return aggregation.getEstimatedSize();
+            return accumulator.getEstimatedSize();
+        }
+    }
+
+    private static class OrderingAccumulator implements Accumulator
+    {
+        private final Accumulator accumulator;
+        private final List<Integer> orderByChannels;
+        private final List<SortOrder> orderings;
+        private final int[] orderByOutputChannels;
+        private final PagesIndex pagesIndex;
+
+        private OrderingAccumulator(
+                Accumulator accumulator,
+                List<Type> orderBySourceTypes,
```

Comment:
consider renaming this to `aggregationSourceTypes` since it's not strictly related to order by but rather aggregation node as a whole